starknet_devnet/blueprints/rpc.py

```diff
@@ -0,0 +1,681 @@
+    accepted_time: int
+    state_diff: RpcStateDiff
+
+
+def rpc_state_update(state_update: dict) -> RpcStateUpdate:
+    """
+    Convert gateway state update to rpc state update
+    """
+    def storage_diffs() -> List[RpcStorageDiff]:
+        _storage_diffs = []
+        for address, storage in state_update["state_diff"]["storage_diffs"].items():
+            diff: RpcStorageDiff = {
+                "address": address,
+                "key": storage["key"],
+                "value": storage["value"]
+            }
+            _storage_diffs.append(diff)
+        return _storage_diffs
+
+    def contracts() -> List[RpcContractDiff]:
+        _contracts = []
+        for contract in state_update["state_diff"]["deployed_contracts"]:
+            diff: RpcContractDiff = {
+                "address": contract["address"],
+                "contract_hash": contract["contract_hash"]
+            }
+            _contracts.append(diff)
+        return _contracts
+
+    rpc_state: RpcStateUpdate = {
+        "block_hash": state_update["block_hash"],
+        "new_root": state_update["new_root"],
+        "old_root": state_update["old_root"],
+        "accepted_time": 0, # FIXME remove hardcoded value
+        "state_diff": {
+            "storage_diffs": storage_diffs(),
+            "contracts": contracts(),
+        }
+    }
+    return rpc_state
+
+
+def rpc_state_diff_contract(contract: dict) -> dict:
+    """
+    Convert gateway contract state diff to rpc contract state diff
+    """
+    return {
+        "address": contract["address"],
+        "contract_hash": f"0x{contract['contract_hash']}",
+    }
+
+
+def rpc_state_diff_storage(contract: dict) -> dict:
+    """
+    Convert gateway storage state diff to rpc storage state diff
+    """
+    return {
+        "address": contract["address"],
+        "key": contract["key"],
+        "value": contract["value"],
+    }
+
+
+class RpcTransaction(TypedDict):
+    """
+    TypedDict for rpc transaction
+    """
+    contract_address: str
+    entry_point_selector: Optional[str]
+    calldata: Optional[List[str]]
+    max_fee: str
+    txn_hash: str
+
+
+def rpc_invoke_transaction(transaction: InvokeSpecificInfo) -> RpcTransaction:
+    """
+    Convert gateway invoke transaction to rpc format
+    """
+    transaction: RpcTransaction = {
+        "contract_address": rpc_felt(transaction.contract_address),
+        "entry_point_selector": rpc_felt(transaction.entry_point_selector),
+        "calldata": [rpc_felt(data) for data in transaction.calldata],
+        "max_fee": rpc_felt(transaction.max_fee),
+        "txn_hash": rpc_felt(transaction.transaction_hash),
+    }
+    return transaction
+
+
+def rpc_deploy_transaction(transaction: DeploySpecificInfo) -> RpcTransaction:
+    """
+    Convert gateway deploy transaction to rpc format
+    """
+    def calldata() -> Optional[List[str]]:
+        # pylint: disable=no-member
+        _calldata = transaction.constructor_calldata
```

Comment:
Deploy transactions don't have `calldata` in gateway but they have `constructor_calldata`. For the rpc purpose I've decided to treat it as calldata, but for example that's no the case in the pathfinder fullnode implementation. 

But that's something RPC standard doesn't really explain.

Reply:
I don't understand; also, why disabling the linting here?

Reply:
I meant that DeployTransactions have `constructor_calldata` instead of `calldata` like InvokeTransactions. I've decided that in RPC I'll treat `constructor_calldata` == `calldata` because it sounds logical to me, but the RPC specs aren't clear on that.

About the disabling lint, now I'm not sure, I'll check if it passes lints without disable and reply here/remove that.

Reply:
I've tested this and for some reason pylint doesn't play nice with these nested functions and thinks transaction is a Dict which is not correct. Same in the other places.